File: BinaryTrees/SerializeDeserialize.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root==null) return "X";
        String leftSer = serialize(root.left);
        String rightSer = serialize(root.right);
        return root.val+","+leftSer+","+rightSer;
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String arr[] = data.split(",");
        Queue<String> q = new LinkedList<>();
        for (int i=0;i<arr.length;i++)
            q.add(arr[i]);
        TreeNode root = deserialize(q);
        return root;
    }
    //Decodes from the queue
    public TreeNode deserialize(Queue<String> q){
        if (q.size()==0)
            return null;
        String peek = q.poll();
        if (peek.equals("X"))
            return null;
        TreeNode root = new TreeNode(Integer.parseInt(peek));
        root.left = deserialize(q);
        root.right = deserialize(q);
        return root;
    }
}

File: C++/Binary Trees/Revision_2025/Serialize_Deserialize.py
from queue import Queue
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return ""
        
        tree_str = ""

        q = Queue()
        q.put(root)

        # Doing a kind-off preorder traversal
        # If null on the left / right -> specifically add "#"
        while not q.empty():
            currNode = q.get()
            
            if currNode is None:
                tree_str += "#,"
            else:
                tree_str += str(currNode.val) + ","

                q.put(currNode.left)
                q.put(currNode.right)
        
        return tree_str


    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if not data:
            return None

        tokens = data.split(",")[:-1]
        
        root_val = int(tokens.pop(0))
        root = TreeNode(root_val)

        q = Queue()
        q.put(root)

        while not q.empty():
            node = q.get()

            left_val = tokens.pop(0)
            if left_val != "#":
                left_node = TreeNode(left_val)
                q.put(left_node)
                node.left = left_node
            
            right_val = tokens.pop(0)
            if right_val != "#":
                right_node = TreeNode(right_val)
                q.put(right_node)
                node.right = right_node
        
        return root

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))
